### PR TITLE
Provide access to Client Certificate

### LINKED ISSUE #2150: Provide a way to access the SSL certificate
We use SSL certificates to connect to intranet / internet. To extract the certificate in other frameworks I write a Filter and using it get the certificates, parse them and get necessary user credentials to Authenticate.

So accordingly I created a HttpServerFilter however calling getAttribute ("javax.servlet.request.X509Certificate") gives me back null, unlike other frameworks. Can  you please advise what is the right way to go about extracting certificates.

What I am doing right now is, that in the Filer, casting the HttpRequest to NettHttpRequest and then getting the SSLHandler from the underline pipeline i.e.

nettyHttpRequest.getChannelHandlerContext().pipeline().get(SSLHandler.class)
and then
sslHandler.engine().getSession().getPeerCertificates()

That is a pretty long chain of relationship to traverse, and ends up with all hard dependencies on Netty specific objects.

Once I have certificate, i extract user info and create the UserDetails and add it back to the HttpRequest.setAttribute so I can have access to it in my RestController or DataFetcher. I couldn't use @ThreadLocal as threads executing controller and filter are different. Hence add it back to the httprequest

I tried SessionBased Authentication and adding AuthenticationProvider but it never gets invoked. Clearly I am doing something wrong. What's the best way for me to handle this ?

### PR DESCRIPTION
**Pull request by BjoernAkAManf**
Current limitations / by design problems i noticed:

- When using 'micronaut.ssl.buildSelfSigned' the option 'micronaut.ssl.clientAuthentication' will be ignored.
- Could not get a p12 Trust Store to work
- Client and Server use same Trust Store in tests

Fixes #2150

### CODE REVIEW

**Comment by jameskleeh** [comment]
I don't understand the purpose of `SSLCertificateProviderHandler`. Seems it only adds a channel attribute which is then read by another handler. Why not do the work in the downstream handler?

**Comment by BjoernAkAManf** [comment]
Well the handler needs to both handle the event provided by the ssl engine and access the corresponding http session. which is why it will add an additional handler.

It might also work by handling that event downstream, but i havent tested that. Is thst what you mean? Id be happy to add that there instead, if you feel that functionality to be more fitting and functional.
I just felt like it doesnt make sense to add such handler in case no ssl has been set and adding a check twice felt ugly aswell.

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/ssl/SSLCertificateProviderHandler.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    private static class AddCertificateToRequest extends ChannelInboundHandlerAdapter {
+        @Override
+        public void channelRead(final ChannelHandlerContext ctx, final Object msg) throws Exception {
+            if (msg instanceof HttpMessage) {
+                HttpMessage<?> request = (HttpMessage<?>) msg;
+                request.setAttribute(CERT_KEY, ctx.channel().attr(CERT_ATTRIBUTE).get());
+                super.channelRead(ctx, msg);
+                return;
+            }
+            throw new UnsupportedOperationException("Message must implement HttpMessage in order to set Certificate!");
```
Comment (jameskleeh) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/ssl/SSLCertificateProviderHandler.java]: Is this necessary? If the message isn't an HttpMessage we can simply ignore it correct?

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/ssl/SSLCertificateProviderHandler.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * Saves a given {@link Certificate} in a {@link HttpMessage}.
+     */
+    private static class AddCertificateToRequest extends ChannelInboundHandlerAdapter {
+        @Override
+        public void channelRead(final ChannelHandlerContext ctx, final Object msg) throws Exception {
+            if (msg instanceof HttpMessage) {
+                HttpMessage<?> request = (HttpMessage<?>) msg;
+                request.setAttribute(CERT_KEY, ctx.channel().attr(CERT_ATTRIBUTE).get());
```
Comment (jameskleeh) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/ssl/SSLCertificateProviderHandler.java]: If the attribute on the channel is null we should probably remove the request attribute here

File: http/src/main/java/io/micronaut/http/HttpRequest.java [hunk 3]
```diff
@@ -149,6 +150,14 @@ default boolean isSecure() {
             .map(Locale::forLanguageTag);
     }
 
+    /**
+     * Retrieves the Certificate used for mutual authentication.
+     *
+     * @return A certificate used for authentication, if applicable.
+     */
+    default Optional<Certificate> getCertificate() {
+        return this.getAttribute("javax.servlet.request.X509Certificate", Certificate.class);
```
Comment (jameskleeh) [inline http/src/main/java/io/micronaut/http/HttpRequest.java]: Create a static variable for this in `HttpAttributes`

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/ssl/SSLCertificateProviderHandler.java (hunk 1 above)
Comment (BjoernAkAManf) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/ssl/SSLCertificateProviderHandler.java]: We could, but that might cause an error thats hard to detect. Maybe log a warning instead?

Comment (BjoernAkAManf) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/ssl/SSLCertificateProviderHandler.java]: I mean it should be an instance of HttpMessage due to the registration anyway, or am i wrong?

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/ssl/SSLCertificateProviderHandler.java (hunk 2 above)
Comment (BjoernAkAManf) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/ssl/SSLCertificateProviderHandler.java]: Not sure if this is redundant though. Shouldn't it be unset anyway?

**Comment by BjoernAkAManf** [comment]
I tried coming up with a solution that utilizes just one only SSLCertificateProviderHandler but i don't think it's that straighforward considering it's registered so early.

Open for suggestions though.

**Pull request merged**